python/ray/tests/test_node_provider_availability_tracker.py

```diff
@@ -0,0 +1,249 @@
+            ),
+        ),
+        "my-node-b": NodeAvailabilityRecord(
+            node_type="my-node-b",
+            is_available=True,
+            last_checked_timestamp=1,
+            unavailable_node_information=None,
+        ),
+    }
+
+    assert summary.node_availabilities == expected
+
+
+def get_timestamp(hour: int, minute: int, second: int, microsecond: int) -> float:
+    # Year, month, day don't just need to be filled out consistently.
+    dt = datetime.datetime(
+        year=2012,
+        month=12,
+        day=21,
+        hour=hour,
+        minute=minute,
+        second=second,
+        microsecond=microsecond,
+    )
+    return dt.timestamp()
+
+
+def test_summary_string(tracker: NodeProviderAvailabilityTracker):
+    global cur_time
+
+    cur_time = get_timestamp(
+        hour=12,
+        minute=13,
+        second=17,
+        microsecond=1234,
+    )
+    exc = NodeLaunchException(
+        "InsufficientInstanceCapacity", "Some message about spot instances."
+    )
+    tracker.update_node_availability("spot-gpu", cur_time, exc)
+
+    cur_time = get_timestamp(
+        hour=12,
+        minute=13,
+        second=51,
+        microsecond=4123,
+    )
+    exc = NodeLaunchException(
+        "InstanceLimitExceeded", "Some message about pesky quotas."
+    )
+    tracker.update_node_availability("newer-gpu-type", cur_time, exc)
+
+    cur_time = get_timestamp(
+        hour=12,
+        minute=14,
+        second=18,
+        microsecond=1,
+    )
+    tracker.update_node_availability("on-demand-gpu", cur_time, None)
+
+    summary = tracker.summary()
+
+    summary_str = summary.summary_string(separator_len=60)
+
+    expected = """
+Launches
+------------------------------------------------------------
+Node types:
+ spot-gpu (attempted=12:13:17): InsufficientInstanceCapacity
+ newer-gpu-type (attempted=12:13:51): InstanceLimitExceeded
+ on-demand-gpu (attempted=12:14:18): Available
+    """.strip()
+
+    assert summary_str == expected
+
+
+def test_summary_from_dict():
+    orig = NodeAvailabilitySummary(
+        node_availabilities={
+            "my-node-a": NodeAvailabilityRecord(
+                node_type="my-node-a",
+                is_available=False,
+                last_checked_timestamp=0,
+                unavailable_node_information=UnavailableNodeInformation(
+                    category="DontFeelLikeIt",
+                    description="This seems like a lot of work.",
+                ),
+            ),
+            "my-node-b": NodeAvailabilityRecord(
+                node_type="my-node-b",
+                is_available=True,
+                last_checked_timestamp=1,
+                unavailable_node_information=None,
+            ),
+        }
+    )
+
+    to_dict = dataclasses.asdict(orig)
+
+    remarshalled = NodeAvailabilitySummary.from_fields(**to_dict)
+
+    assert orig == remarshalled
+
+
+if __name__ == "__main__":
+    if os.environ.get("PARALLEL_CI"):
```

Comment:
what's this

Reply:
idk lol it's in all of the test files now (i checked `test_basic` and `test_autoscaler`)

Reply:
Oh, ok then.